packages/typescript-private/device-server/src/device-host/device-host.download-shared-resource.ts

```diff
@@ -0,0 +1,165 @@
+import { DefaultHttpOptions, DuplicatedCallGuarder, Instance, stringifyError } from '@dogu-tech/common';
+import { DeviceHostDownloadSharedResource } from '@dogu-tech/device-client-common';
+import { HostPaths, TaskQueue, TaskQueueListener, TaskQueueTask } from '@dogu-tech/node';
+import { Injectable } from '@nestjs/common';
+import { OnEvent } from '@nestjs/event-emitter';
+import axios from 'axios';
+import fs from 'fs';
+import path from 'path';
+import stream, { Stream } from 'stream';
+import { v4 as uuidv4 } from 'uuid';
+import { OnUpdateEvent } from '../events';
+import { DoguLogger } from '../logger/logger';
+
+export type DeviceHostDownloadResult = Instance<typeof DeviceHostDownloadSharedResource.receiveMessage> & { message: string };
+export type DeviceHostDownloadParam = Instance<typeof DeviceHostDownloadSharedResource.sendMessage>;
+
+class DeviceHostDownloadTask extends TaskQueueTask<DeviceHostDownloadResult> {
+  constructor(public readonly param: DeviceHostDownloadParam, task: () => Promise<DeviceHostDownloadResult>, listeners: TaskQueueListener<DeviceHostDownloadResult>[] = []) {
+    super(task, listeners);
+  }
+}
+
+type DeviceHostDownloadTaskQueue = TaskQueue<DeviceHostDownloadResult, DeviceHostDownloadTask>;
+interface DeviceHostDownloadTaskRunningQueue {
+  filePath: string;
+  updateGuarder: DuplicatedCallGuarder;
+  queue: DeviceHostDownloadTaskQueue;
+}
+
+@Injectable()
+export class DeviceHostDownloadSharedResourceService {
+  constructor(private readonly logger: DoguLogger) {}
+  private dividerQueue: DeviceHostDownloadTaskQueue = new TaskQueue();
+  private runningQueues: DeviceHostDownloadTaskRunningQueue[] = [];
+  private onUpdateGuarder = new DuplicatedCallGuarder();
+
+  public async queueDownload(message: DeviceHostDownloadParam): Promise<DeviceHostDownloadResult> {
+    const task = new DeviceHostDownloadTask(message, async () => {
+      return await this.download(message);
+    });
+    const result = await this.dividerQueue.scheduleAndWait(task);
+    if (!result.success) {
+      throw result.error;
+    }
+    return result.value;
+  }
+
+  @OnEvent(OnUpdateEvent.key)
+  async onUpdate(value: Instance<typeof OnUpdateEvent.value>): Promise<void> {
+    await this.onUpdateGuarder.guard(() => {
```

Comment:
wrap try {} catch {} or append .catch() because onUpdate if rejected then does not called again